File: python/agent/trace.py
"""Per-session JSONL diagnostics without credentials or hidden reasoning."""

import asyncio
import hashlib
from datetime import datetime, timezone
from pathlib import Path

from .jsonutil import dumps


def timestamp():
    return datetime.now(timezone.utc).isoformat(timespec="milliseconds").replace("+00:00", "Z")


def preview(value, limit=800):
    try:
        text = value if isinstance(value, str) else dumps(value)
    except (TypeError, ValueError):
        text = "[unserializable]"
    return text if len(text) <= limit else text[:limit] + " … [truncated]"


class TraceWriter:
    def __init__(self, data_dir):
        self.directory = Path(data_dir).resolve() / "traces"

    async def write(self, user_id, session_id, events):
        def append():
            self.directory.mkdir(parents=True, exist_ok=True)
            key = "-".join(hashlib.sha256(x.encode()).hexdigest() for x in (user_id, session_id))
            with (self.directory / f"{key}.jsonl").open("a", encoding="utf-8") as stream:
                stream.write("".join(dumps(event) + "\n" for event in events))
        await asyncio.to_thread(append)

File: src/trace.js
import { mkdir, appendFile } from 'node:fs/promises';
import { join } from 'node:path';
import { createHash } from 'node:crypto';

export class TraceWriter {
  constructor(dataDir) { this.dir = join(dataDir, 'traces'); }
  async write(userId, sessionId, events) {
    await mkdir(this.dir, { recursive: true });
    const name = createHash('sha256').update(JSON.stringify([userId, sessionId])).digest('hex');
    await appendFile(join(this.dir, `${name}.jsonl`), events.map(e => JSON.stringify(e)).join('\n') + '\n', { mode: 0o600 });
  }
}

export function preview(value, limit = 1200) {
  const str = typeof value === 'string' ? value : (JSON.stringify(value) ?? 'null');
  return str.length > limit ? str.slice(0, limit) + '…[truncated]' : str;
}
